tools/approval-service/internal/approvalservice/eventprocessor.go

```diff
@@ -0,0 +1,169 @@
+	"github.com/gravitational/teleport/api/types"
+)
+
+type processor struct {
+	// TeleportUser is the user that the approval service will use to request access.
+	TeleportUser string
+	// TeleportRole is the role that the approval service will request access to.
+	TeleportRole string
+
+	teleportClient teleClient
+	githubClient   ghClient
+	log            *slog.Logger
+
+	// envNameToID is a cache of environment names to their IDs.
+	// This is expected to be read-only after Setup() is called.
+	// If this is not the case, it will need to be protected by a mutex.
+	envNameToID map[string]int64
+	// validation is used to populate envNameToID during setup.
+	validation []config.Validation
+}
+
+// Small interfaces around the Teleport and GitHub clients to keep implementation details separate.
+type teleClient interface {
+	CreateAccessRequestV2(ctx context.Context, req types.AccessRequest) (types.AccessRequest, error)
+}
+type ghClient interface {
+	UpdatePendingDeployment(ctx context.Context, info github.PendingDeploymentInfo) ([]github.Deployment, error)
+	GetEnvironment(ctx context.Context, info github.GetEnvironmentInfo) (github.Environment, error)
+}
+
+var _ teleClient = &teleportClient.Client{}
+var _ ghClient = &github.Client{}
+
+func newProcessor(cfg config.Root, ghClient ghClient, teleClient teleClient) *processor {
+	return &processor{
+		TeleportUser: cfg.Teleport.User,
+		TeleportRole: cfg.Teleport.RoleToRequest,
+
+		teleportClient: teleClient,
+		githubClient:   ghClient,
+		validation:     cfg.GitHubEvents.Validation,
+		envNameToID:    make(map[string]int64),
+		log:            slog.Default(),
+	}
+}
+
+func (p *processor) Setup() error {
+	if len(p.validation) == 0 {
+		return fmt.Errorf("no environments configured")
+	}
+	// Get environment IDs for each environment we support
+	for _, v := range p.validation {
+		for _, env := range v.Environments {
+			env, err := p.githubClient.GetEnvironment(context.TODO(), github.GetEnvironmentInfo{
+				Org:         v.Org,
+				Repo:        v.Repo,
+				Environment: env,
+			})
+			if err != nil {
+				return fmt.Errorf("getting environment %q: %w", env, err)
+			}
+			p.envNameToID[envEntry(v.Org, v.Repo, env.Name)] = env.ID
+		}
+	}
+
+	return nil
+}
+
+func (p *processor) ProcessDeploymentReviewEvent(e githubevents.DeploymentReviewEvent, valid bool) error {
+	if !valid {
+		// TODO: Create a rejected access request if the event is invalid (e.g. incorrect org, env, user, etc.)
+		//       This should be useful for audit purposes.
+		return fmt.Errorf("invalid event")
+	}
+
+	// Create access request
+	// The name of an access request must be a UUID
+	name := uuid.New().String()
+	p.log.Info("Creating access request", "access_request_name", name)
+	req, err := types.NewAccessRequest(name, p.TeleportUser, p.TeleportRole)
+	if err != nil {
+		return fmt.Errorf("generating access request: %w", err)
+	}
+
+	req.SetRequestReason("Deployment review for environment " + e.Environment)
```

Comment:
This will need workflow call parameters as well, which means accepting another webhook, and might require waiting for the webhook call before filing and access request. Alternatively, if the access request can be modified, the workflow call parameters could update the request when the webhook call is made.